File: x86_64/updatereadme.py
import glob
import subprocess
import time
import os
import io
import json
import math
import shutil
import zstandard
from datetime import datetime as dt


def suffix(d):
    return "th" if 11 <= d <= 13 else {1: "st", 2: "nd", 3: "rd"}.get(d % 10, "th")


def custom_strftime(format, t):
    return t.strftime(format).replace("{d}", str(t.day) + suffix(t.day))


datetime = custom_strftime("%a {d}, %b %Y at %I:%M:%S%p", dt.now())


def filebrowser(ext=""):
    "Returns files with an extension"
    return [file for file in glob.glob(f"*{ext}")]


files = filebrowser(".pkg.tar.zst")

files.sort()

pkgcount = 0
for file in files:
    pkgcount += 1
print(f"There are {pkgcount} packages to be uploaded!\n")

home = os.path.expanduser("~")
readme = open("../README.md", "w")
installme = open(home + "/.config/package-list", "w")
data = []
readme.write(
    f"# <img src='favicon.svg' width='64' height='64'> Arch Linux Repo <img src='favicon.svg' width='64' height='64'>\n"
)
badges = f"\n<p align='center'>\n\
  <img src='https://img.shields.io/badge/Maintained-Yes-green?colorA=434c5e&colorB=ff59f9&style=flat-square'>\n\
  <img src='https://img.shields.io/github/actions/workflow/status/HeCodes2Much/archlinux.repo.063240.xyz/pages/pages-build-deployment?colorA=434c5e&colorB=ff59f9&style=flat-square'>\n\
  <img src='https://img.shields.io/github/last-commit/HeCodes2Much/archlinux.repo.063240.xyz?colorA=434c5e&colorB=ff59f9&style=flat-square'>\n\
  <img src='https://img.shields.io/github/repo-size/HeCodes2Much/archlinux.repo.063240.xyz?colorA=434c5e&colorB=ff59f9&style=flat-square'>\n\
  <img src='https://img.shields.io/static/v1?label=Packages&message={pkgcount}&colorA=434c5e&colorB=ff59f9&style=flat-square'>\n\
  <img src='https://img.shields.io/github/license/HeCodes2Much/archlinux.repo.063240.xyz?colorA=434c5e&colorB=ff59f9&style=flat-square'>\n\
  <img src='https://img.shields.io/github/issues/HeCodes2Much/archlinux.repo.063240.xyz?colorA=434c5e&colorB=ff59f9&style=flat-square'>\n\
  <img src='https://img.shields.io/github/stars/HeCodes2Much/archlinux.repo.063240.xyz?colorA=434c5e&colorB=ff59f9&style=flat-square'>\n\
  <img src='https://img.shields.io/github/forks/HeCodes2Much/archlinux.repo.063240.xyz?colorA=434c5e&colorB=ff59f9&style=flat-square'>\n\
  <img src='https://img.shields.io/github/commit-activity/m/HeCodes2Much/archlinux.repo.063240.xyz?colorA=434c5e&colorB=ff59f9&style=flat-square'>\n\
</p>\n"

readme.write(badges)
readme.write(f"\n## Software\n")


def get_file_name(file):
    global name
    head = "head -n5"
    awk = "awk '{$1=$2=\"\"; print $0}'"

    try:
        with open(file, "rb") as fh:
            dctx = zstandard.ZstdDecompressor(max_window_size=2147483648)
            stream_reader = dctx.stream_reader(fh)
            text_stream = io.TextIOWrapper(stream_reader, encoding="utf-8")
            for line in text_stream:
                if line.startswith("pkgname"):
                    command = (
                        f"echo '{line}' | {head} | grep -I pkgname | {awk} | sed -n 1p"
                    )
                    process = subprocess.Popen(
                        command, stdout=subprocess.PIPE, stderr=None, shell=True
                    )
                    output = process.communicate()
                    name = str(output[0].decode()).strip()
    except UnicodeDecodeError:
        pass
    except UnboundLocalError:
        pass

    if not name:
        get_file_name(file)
    else:
        return name


def get_file_version(file):
    global version
    head = "head -n5"
    awk = "awk '{$1=$2=\"\"; print $0}'"
    try:
        with open(file, "rb") as fh:
            dctx = zstandard.ZstdDecompressor(max_window_size=2147483648)
            stream_reader = dctx.stream_reader(fh)
            text_stream = io.TextIOWrapper(stream_reader, encoding="utf-8")
            for line in text_stream:
                if line.startswith("pkgver"):
                    command = (
                        f"echo '{line}' | {head} | grep -I pkgver | {awk} | sed -n 1p"
                    )
                    process = subprocess.Popen(
                        command, stdout=subprocess.PIPE, stderr=None, shell=True
                    )
                    output = process.communicate()
                    version = str(output[0].decode()).strip()
    except UnicodeDecodeError:
        pass
    except UnboundLocalError:
        pass

    if not version:
        get_file_version(file)
    else:
        return version


def get_file_date(file):
    file_date = os.path.getmtime(file)
    return str(time.strftime("%H:%M:%S %d-%m-%Y", time.localtime(file_date)))


def convert_size(size_bytes):
    if size_bytes == 0:
        return "0B"
    size_name = ("B", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB")
    i = int(math.floor(math.log(size_bytes, 1024)))
    p = math.pow(1024, i)
    s = round(size_bytes / p, 2)
    return str("%s %s" % (s, size_name[i]))


def get_file_size(filename):
    st = os.stat(filename)
    return str(convert_size(st.st_size))


for file in files:
    name = str(get_file_name(file))
    version = str(get_file_version(file))
    size = str(get_file_size(file))
    date = str(get_file_date(file))

    print(f"File Updated: Name ({name}), Version ({version}) Size {size} Date {date}")

    userpac = "<p> <span class='red'>[</span> <span class='yellow'>arch</span><span class='green'>@</span><span class='purple'>linuxrepos.org</span> <span class='red'>~]</span> <br> <span class='green'>pacman</span> <span class='orange'>-Si</span>"
    pacgrep = "<span class='purple'>|</span> <span class='green'>grep</span> <span class='yellow'>'Name\|Version\|Installed\|Build'</span> <span class='purple'>|</span> <span class='green'>awk</span> <span class='orange'>-F</span><span class='yellow'>':' '{print $2}'<span> <span class='purple'>|</span> <span class='green'>sed</span> <span class='yellow'>':a;N;$!ba;s/\\n/ |/g'</span><br>"

    readme.write(
        f"{userpac} <span class='foreground'><a href='docs/{name}'>{name}</a></span> {pacgrep} <span class='foreground'><a href='docs/{name}'>{name} \t|\t {version} \t|\t {size} \t|\t {date}</a></span> </p>\n"
    )

    installme.write(f"{name}\n")
    data.append(f"{name}")

multiline_addrepo = (
    f"\n## Add my repo\n"
    f"* **Maintainer:** [HeCodes2Much](https://aur.archlinux.org/account/HeCodes2Much/)\n"
    f"* **Description:**  A repository with some AUR packages that the team uses\n"
    f"* **Upstream page:** https://archlinux.repo.063240.xyz/\n"
    f"* **Key-ID:** 75A3 8DC6 84F1 A0B8 0891  8BCE E30E C2FB FB05 C44F \n"
    f"* **Fingerprint:** [download](http://pgp.net.nz:11371/pks/lookup?op=vindex&fingerprint=on&search=0xE30EC2FBFB05C44F)\n"
    f"\nAppend to */etc/pacman.conf*:\n```\n[linuxrepos]\nSigLevel = Required DatabaseOptional\nServer = https://archlinux.repo.063240.xyz/$arch/\n```"
    f"\nTo check signature, add my key:\n"
    f"```\nsudo pacman-key --keyserver hkp://pgp.net.nz --recv-key 75A38DC684F1A0B808918BCEE30EC2FBFB05C44F\nsudo pacman-key --keyserver hkp://pgp.net.nz --lsign-key 75A38DC684F1A0B808918BCEE30EC2FBFB05C44F\n```"
    # f"\nYou may also want to \nAppend to */etc/pacman.conf*:\n```\n[linuxrepos-git]\nSigLevel = Optional TrustAll\nServer = https://github.com/HeCodes2Much/archlinux.repo.063240.xyz/releases/download/$repo/\n```"
)

readme.write(multiline_addrepo)

multiline_showsupport = (
    f"\n## Show your support\n"
    f"\nGive a ⭐️ if this project helped you!\n"
    f"\nThis README was generated with ❤️ by [HeCodes2Much](https://github.com/HeCodes2Much/)\n"
    f"*   Last updated on: {datetime}\n"
)

readme.write(multiline_showsupport)

readme.close()
shutil.copyfile("../README.md", "../repo.md")

json_string = json.dumps(data, indent=4)
with open("../packages.json", "w") as outfile:
    outfile.write(json_string)

File: x86_64/updatereadme.rb
# coding: utf-8
require 'date'
require 'active_support/core_ext/integer/inflections'
require 'open3'
require 'open-uri'
require 'json'

def get_file_name(file)
  begin
    awk1 = "awk '{print $0}'"
    awk2 = "awk '{$1=$2=\"\"; print $0}'"
    command = "bsdtar -xOf #{file} | #{awk1} | grep -I pkgname | #{awk2}"
    stdin, stdout, stderr, wait_thr = Open3.popen3(command)
    output = stdout.gets(nil)
    return output.strip.to_s
  rescue => e
    retry
  end
end

def get_file_version(file)
  begin
    awk1 = "awk '{print $0}'"
    awk2 = "awk '{$1=$2=\"\"; print $0}'"
    command = "bsdtar -xOf #{file} | #{awk1} | grep -I pkgver | #{awk2}"
    stdin, stdout, stderr, wait_thr = Open3.popen3(command)
    output = stdout.gets(nil)
    return output.strip.to_s
  rescue => e
    retry
  end
end

def get_aur_name(name)
  begin
    packages = []
    response = URI.open("https://aur.archlinux.org/rpc/?v=5&type=search&arg=#{CGI.escape(name)}").read
    data = JSON.parse(response)
    
    title = data["results"]
    title.each do |title|
      new_name = title["Name"]
      packages.push(new_name)
    end

    if packages.include?("#{name}")
      return "#{name}"
    elsif packages.include?("#{name}-git")
      return "#{name}-git"
    elsif packages.include?("#{name}-bin")
      return "#{name}-bin"
    else
      return "#{name}"
    end
  rescue => e
    retry
  end
end

open("../README.md", "w") { |f|
  f.write("")
}
home = Dir.home
open("#{home}/.config/package-list", "w") { |f|
  f.write("")
}

badges = ("# <img src='favicon.ico'> The Repo Club's Arch Repo <img src='favicon.ico'>\n
<p align='center'>\n\
  <img src='https://img.shields.io/badge/Maintained-Yes-green?color=red&style=flat-square'>\n\
  <img src='https://img.shields.io/github/last-commit/TheCynicalTeam/Arch.TheRepo.Club?color=red&style=flat-square'>\n\
  <img src='https://img.shields.io/github/repo-size/TheCynicalTeam/Arch.TheRepo.Club?color=red&style=flat-square'>\n\
  <img src='https://img.shields.io/static/v1?label=License&message=CC%20BY-NC-SA%204.0&color=red&style=flat-square'>\n\
  <img src='https://img.shields.io/github/issues/TheCynicalTeam/Arch.TheRepo.Club?color=red&style=flat-square'>\n\
  <img src='https://img.shields.io/github/stars/TheCynicalTeam/Arch.TheRepo.Club?color=red&style=flat-square'>\n\
  <img src='https://img.shields.io/github/forks/TheCynicalTeam/Arch.TheRepo.Club?color=red&style=flat-square'>\n\
  <img src='https://img.shields.io/github/commit-activity/m/TheCynicalTeam/Arch.TheRepo.Club?color=red&style=flat-square'>\n\
</p>
\n## Software\n")

open("../README.md", "a") { |f|
  f.write(badges)
}

files = Dir["*.pkg.tar.zst"].sort
for file in files
  begin
    ignore = open("../ignorepackages", "r").read
    name = get_file_name(file.to_s)
    if not ignore.include?(name)
      version = get_file_version(file.to_s)
      if not name
        name = get_file_name(file.to_s)
      end
      if not version
        version = get_file_name(file.to_s)
      end

      print("File Updated: (#{name} v#{version})\n")
      aur_name = "#{get_aur_name(name)}"
      open("../README.md", "a") { |f|
        f.write("*   [#{name}](docs/#{name}/) Version: #{version} ![AUR maintainer](https://img.shields.io/aur/maintainer/#{aur_name}?color=blue&style=flat-square) ![AUR maintainer](https://img.shields.io/aur/license/#{aur_name}?color=orange&style=flat-square)\n")
      }
      open("#{home}/.config/package-list", "a") { |f|
        f.write("#{name} #{version}\n")
      }
    end
  rescue TypeError
    retry
  end
end

multiline_addrepo = ("\n## Add my repo\n"\
"* **Maintainer:** [TheCynicalTeam](https://aur.archlinux.org/account/TheCynicalTeam/)\n"\
"* **Description:**  A repository with some AUR packages that the team uses\n"\
"* **Upstream page:** https://arch.therepo.club/\n"\
"* **Key-ID:** 10DF 44AC D4C8 4539 53B7 CCBA 206A DED6 6160 901B\n"\
"* **Fingerprint:** [download](http://pgp.net.nz:11371/pks/lookup?op=vindex&fingerprint=on&search=0x96414492E2220753)\n"\
"\nAppend to */etc/pacman.conf*:\n```\n[therepoclub]\nSigLevel = Required DatabaseOptional\nServer = https://arch.therepo.club/$arch/\n```"\
"\nTo check signature, add my key:\n"\
"```\nsudo pacman-key --keyserver hkp://pgp.net.nz --recv-key 75A38DC684F1A0B808918BCEE30EC2FBFB05C44F\nsudo pacman-key --keyserver hkp://pgp.net.nz --lsign-key 75A38DC684F1A0B808918BCEE30EC2FBFB05C44F\n```")

open("../README.md", "a") { |f|
  f.write(multiline_addrepo)
}

d = Time.new
dt = DateTime.now.strftime("Last updated on: %a #{d.day.ordinalize}, %b %Y at %I:%M:%S%p")

multiline_showsupport = ("\n## Show your support\n"\
"\nGive a ⭐️ if this project helped you!\n"\
"\nThis README was generated with ❤️ by [TheCynicalTeam](https://github.com/TheCynicalTeam/)\n"\
"*   #{dt}")

open("../README.md", "a") { |f|
  f.write(multiline_showsupport)
}
